### ISSUE TITLE
additional_dependencies for system language hooks

### ISSUE THREAD

**Issue by marcjay** [pre-commit/pre-commit#1453]
I'm aware of the requirement that additional dependencies for system hooks - the 'escape hatch' for hooks that don't fit into the python/node/go etc. managed environments list, such as shellcheck, must be installed outside of pre-commit in order to work.

Also aware of the python module they can be used to 'mirror' shellcheck in particular, though that feels a little hacky.

I'm sure you've given this thought before, and there's a very good reason why this hasn't been done, but, is it conceivable, with a PR I'd be willing to write, that additional dependencies could be supported for non-python etc. hooks?

I'm thinking along the lines of being able to specify a URL to a ZIP, which pre-commit would `wget` into its cache directory, sorting a checksum so it could 'manage' when to re-download it, and then adding the directory to the front of `$PATH` when running the hook.

I'm sure it's not possible, but would be keen to know why, as there are a few hooks I use that this applies too and would feel like a useful addition for others.

**Comment by asottile** [comment]
I don't think that would be a feature to `language: system` -- I think that would be a separate `language` itself (maybe `language: download` or something).  in the ideal world, pre-commit would learn how to install environments of all major programming languages and then you'd be able to install the tool you're looking for in that manner.  there's [directions for adding a new language here](https://github.com/pre-commit/pre-commit/blob/master/CONTRIBUTING.md#adding-support-for-a-new-hook-language) and several languages have been contributed by external contributors already!

also `python` is not the only language which supports `additional_dependencies`, `go`, `ruby`, `node`, and a few others also support the setting (and presumably languages which are added can also support that as well).

for shellcheck in particular, [shellcheck-py](https://github.com/shellcheck-py/shellcheck-py) is the suggested tool for pre-commit

note that a url alone does not help you here since you need to work about platform and architecture and likely checksumming, etc.

for `system` / `script` hooks in particular, they do not get any "environment" to install things into (they're run directly from configuration-only)

**Comment by marcjay** [comment]
> I think that would be a separate `language` itself (maybe `language: download` or something)

That's a good point, yup

> in the ideal world, pre-commit would learn how to install environments of all major programming languages and then you'd be able to install the tool you're looking for in that manner.

Is always installing from source ideal? The language a hook is written in feels like it should be an implementation detail as much as possible. We don't use Go at all for instance, but if we want to use a hook written in Go, using `language: golang` we need to have a working Go toolchain installed locally. (I did see your comment about it being a second-class language at the moment, and that it's intended to become a first class one - https://github.com/pre-commit/pre-commit/issues/1449#issuecomment-626923117). Shellcheck, amongst some other tools, is written in Haskell - I'm not sure I like the idea of needing a Haskell toolchain installed locally. Again, I know about shellcheck-py, but needing to use a wrapper that relies on a third party keeping it up to date with new releases of shellcheck seems sub-optimal.

Appreciate the need to handle platforms, architectures and checksums, but being able to point at a pre-built binary, pull it and run it seems useful?

**Comment by asottile** [comment]
yeah I'm not saying `language: download` is off the table, just that it's not as simple as you're making it out to be

**Comment by marcjay** [comment]
> I'm sure you've given this thought before, and there's a very good reason why this hasn't been done

> I'm sure it's not possible

I don't think it's that simple - I don't think I could have added any more caveats up top 😄 

I guess I just wanted to know if 
a) there's a known reason why it's not possible/sensible
b) if a PR was put out for it, would it be rejected as not something you'd want in the project

If either are true, I wasn't going to expend any energy having a deeper look at it. On the basis that it seems to get a luke warm reception, I'll have a proper look. And hey, worst case, I'll run into a blocker and leave a comment here to save the next person from wasting time on it :)

**Comment by asottile** [comment]
yeah nothing has been attempted there

**Comment by asottile** [comment]
would you like to design / work on such a thing?

**Comment by marcjay** [comment]
Yes I would like to. I'm aiming to look at it in the next day or so

**Comment by marcjay** [comment]
Timeline was ambitious but not forgotten about. Made a start at the time but need to find some more time to pick it up again. If you'd rather have a cleaner issues page, by all means close this and I can come back when/if there's a PR to be had

**Comment by marcjay** [comment]
I've been working on this further and noticed a general case which might be simpler, and cover most of what I was looking for in the first place. It's 50/50 on whether you'll hate the idea, if you don't, I'll raise a separate issue to track it and work on it.

Premise 1: A lot of tools that aren't pip installable are on GitHub, and when they are released, artifacts are available from the releases page, with artifacts built for most OS + arch combinations
Premise 2: These artifacts have names that include the OS + arch in some way that can be pattern-matched (like `...linux_x86_68...`, `...darwin-amd64...`)

Packaging aside (zip vs tar.xz vs bare), this is true for a number of repos - some examples: 
https://github.com/koalaman/shellcheck/releases
https://github.com/mvdan/sh/releases
https://github.com/terraform-docs/terraform-docs/releases
https://github.com/terraform-linters/tflint/releases

What do you think to the following rough proposal for config?:
`.pre-commit-config.yaml`:
```
repos:
  - repo: https://github.com/terraform-linters/tflint
    rev: v0.19.0
    hooks:
      - id: tflint
```
`.pre-commit-hooks.yaml`:
```
- id: tflint
  name: Terraform Lint
  description: TFLint is a Terraform linter focused on possible errors, best practices, and so on.
  language: github-release
  entry: tflint
  types: [terraform]
```

`language: github-release` would work by:
1. Calling the GitHub [Get a release by tag name](https://docs.github.com/en/rest/reference/repos#get-a-release-by-tag-name) API (no authentication needed) to get the list of assets for the `rev` (yes I admit there is a limitation this must be a tag and not a sha). In the above example: https://api.github.com/repos/terraform-linters/tflint/releases/tags/v0.19.0
1. Find the asset that matches the current OS + arch
1. Download the matching asset, unpack it if needed and place it on the path (no checksumming should be needed I believe, given the relationship with `rev`?)

With the benefits as before, that you don't need to have the toolchain installed for whatever language the hook was written in, and there's a good chance it can download and unpack an asset faster than it can build from source on that initial environment setup.

Is that a terrible idea?

**Comment by marcjay** [comment]
Before I tidy this up and figure out how to write tests for it all, could I run the (working) proposal past you for `language: download`?
(I noted the `:` separator in `cli:{package_name}:{package_version}` for [rust](https://pre-commit.com/#rust) but this gets messy here given the `:` in URIs, so opted for pipes)

syntax: `<os>|<arch>|<url>('->'<name>)(|<checksum>)`

Example (checksum is optional - provided in the darwin case):
```
additional_dependencies: [
  'darwin|x86_64|https://github.com/koalaman/shellcheck/releases/download/v0.7.1/shellcheck-v0.7.1.darwin.x86_64.tar.xz|b080c3b659f7286e27004aa33759664d91e15ef2498ac709a452445d47e3ac23',
  'linux|x86_64|https://github.com/koalaman/shellcheck/releases/download/v0.7.1/shellcheck-v0.7.1.linux.x86_64.tar.xz'
]
```

`shfmt` is an example of an unpackaged artifact, so an optional `->NAME` can be used to transform the downloaded binary to a sensible version-independent name:
```
additional_dependencies: [
  'darwin|x86_64|https://github.com/mvdan/sh/releases/download/v3.1.2/shfmt_v3.1.2_darwin_amd64->shfmt',
  'linux|x86_64|https://github.com/mvdan/sh/releases/download/v3.1.2/shfmt_v3.1.2_linux_amd64->shfmt'
]
```

Keen to know how close this is to what you're expecting/what needs to change. Cheers!

**Comment by asottile** [comment]
(copying from discord)

> **asottile**: some initial thoughts: checksum should be required, is the platform info enough to capture everything? archives are structured in all sorts of ways how do they get extracted generically (maybe that's fine and entry should capture that? kinda act like script, but instead of joinpath from repo root joinpath from extract-point root?), how do we handle all the archive formats (python only has support for a handful)?
> **marcjay**: Fair enough on mandatory checksums. I went looking at a bunch of projects and their release binaries, and I believe OS + arch captures everything. 
On structure, yeah I think entry is sufficient to handle any structure. In my POC if there is a single top level folder in the archive, it moves everything up a level to get rid of it (you might hate that), but it simplifies the entry if there's a top level directory with the version info etc. in - so you aren't duplicating the version in the URL and the entry
In my searching I only came across zips and tar.gz's - are you thinking of 7zip or something? Maybe support for those two is enough?
> **asottile**: yeah 7z is tricky from a licensing standpoint

**Comment by skorokithakis** [comment]
My comment is going to be not-very-related to this, but it aims to solve the same problem:

I have the binaries committed in the repo (fine by me for now, so I can version them), the problem is that `pre-commit` doesn't even *tell me where the repo is*! If there were a `PRE_COMMIT_REPOSITORY` env var, I would manually run `$PRE_COMMIT_REPOSITORY/myfile` as the executable and side-step the lack of being able to download binaries.

I could even use `$PRE_COMMIT_REPOSITORY` as a space to download the executables I needed in to, until this functionality is implemented. Is there a way to do that currently?

**Comment by asottile** [comment]
I assume you're using a `language: script` hook?  you can use `dirname "$0"` (or in python `__file__`, etc.) to figure out where your script lives and access files relative to it

**Comment by skorokithakis** [comment]
I was using a `system` hook (hence the difficulty), but I can switch to Python if that'll help. I didn't because I thought I'd have to publish the script on PyPI first, but it doesn't sound like that's necessary, thanks!

**Comment by asottile** [comment]
you could also use `script` directly -- if your binaries are checked in at `bin/foo` you would use `entry: ./bin/foo`

**Comment by skorokithakis** [comment]
I tried that but it didn't work because the working directory is the target repo; the scripts are checked in to the hook repo.

**Comment by asottile** [comment]
correct, for `language: system`, incorrect for `language: script`

**Comment by skorokithakis** [comment]
Oh oops, I misread, I'll try that, thanks!

**Comment by Skylion007** [comment]
To simplify this further: there is this little helper for installing binary releases from Github: https://github.com/crate-ci/gh-install/
Discussed in this issue: https://github.com/crate-ci/typos/issues/285

**Comment by asottile** [comment]
that doesn't simplify anything, it's very specific to github and not a solution for this more general problem

**Comment by MaxymVlasov** [comment]
Why not just add [asdf](https://github.com/asdf-vm/asdf) or analog? Also, it can install pre-commit too.

For me, that issue sound little bit as out of scope of what pre-commit should do - check files, not manage deps for checks.

I think, it will be more usefull, have integration beetween pre-commit and deps manager tool.

Also, that integration can be implemented just as shell hook that install asdf, run asdf to install deps and then run another hooks

**Comment by asottile** [comment]
nah you're missing the point, the whole point of pre-commit is *it manages your tools*

**Comment by Cyberbeni** [comment]
What is the status on this? From glancing over the thread, I think not even the interface is final?

My idea is to have something like a `download_entry` which gets a target directory and the rev passed into and if the script returns with 0 then the stuff placed into the target directory goes into the cache. We have some vendored tools in a jfrog artifactory and the download script requires a secret that shouldn't be included in the git repository, so just storing URLs for different platforms/architectures in the pre-commit config wouldn't really work for us.

What do you think?

**Comment by asottile** [comment]
the issue is to to date with the current status

**Comment by tokudu** [comment]
@asottile what are your thoughts on supporting `additional_dependencies` for `script` hooks, where it can specify a list of other local setup scripts. These dependency scripts would be invoked only once per hook, or whenever thir files are modified. Something like this would but a lot more generic than dealing with URLs & downloading, where things like different types of authentication quickly become a cascading list of requirements to support. 

Overall, what the framework offers today works well. You can always write a wrapper script that downloads dependencies if they are missing. The only problem is that you have to either disable parallel execution via `require_serial: true` or implement some locking in bash to handle concurrent installations, which becomes pretty annoying to do in a cross-platform way without depending on more tools. 

I really like @alexeagle suggestion, but you still end up with a question: "what if bazel is not installed" :)

**Comment by asottile** [comment]
`script` hooks have no environment so they can't have `additional_dependencies` -- what we're discussing here is a new `language`

**Comment by alexeagle** [comment]
I'm not sure that it's commonly-enough used, but as a simpler middle ground to have Bazel manage hermetic toolchains, you could just have `bazel` be a language? Then pre-commit installs the bazel-version-selection wrapper https://github.com/bazelbuild/bazelisk and everything following that is guaranteed to be exact versions of interpreters/runtimes/libraries etc.

**Comment by asottile** [comment]
sorry but I've dealt with and managed bazel before and I want to stay as far away from that hell as I can

**Comment by dudicoco** [comment]
I really like the idea of using bazel to manage dependencies.
@asottile can you elaborate on what issues you have with bazel?

**Comment by asottile** [comment]
it was hell on earth at stripe. I don't care if you like bazel I am not going anywhere near it again

**Comment by asottile** [comment]
it's also way off topic for this issue so please stop

**Comment by sondrelg** [comment]
Tried writing a tiny Rust pre-commit hook recently and even with just a single (relatively slim) dependency, Rust needed over a minute to pull and compile the hook locally. To me it seems like the ability to pull a pre-built binary instead of building things from scratch would really improve the experience of using those languages for hooks. With that in mind it would be really great to see this ticket gain some traction again.

Attempting to summarize the progress made so far:

### Objectives
- It would be useful to create a new pre-commit language that downloads pre-compiled binaries. This removes the need for a local toolchain and greatly reduces the time it takes to install a hook.
- We need to consider platform and arch (since all platforms and archs will never be supported, I take it that missing combinations should lead to pre-commit just raising a not-found error when trying to install the hook or something similar?)
- It's best to steer clear of depending on vendor specific solutions like Github's releases 

### Issues
- We don't know how to identify binary URLs. In this thread we have considered either:
    - Retrieving binary targets and relying on a naming conventions like  `<binary-name>-<version>-<arch>-<platform>` to identify which binary to use
    - Specifying all targets and their URLs explicitly, like explored [here](https://github.com/pre-commit/pre-commit/issues/1453#issuecomment-685912845), and 

The downsides of each of these seems to be that:
- It's not possible to just retrieve binary targets from a base URL. Even for Github releases, it would be necessary to first depend on the Github releases API before trying to identify which binary to fetch
- specifying targets explicitly (in `additional_dependencies`) seems like it would be decouple binary targets from sha/tag and we would not be able to autoupdate them

------

The main downside of the previously discussed `github_releases` option, except from being vendor specific seems to be that sha would not be meaningful, but after exploring remaining options that now seems like a small price to pay to enable downloading pre-built binaries in general. Does the vendor specificity definitely still rule this option out @asottile? If not, a Github releases language seems like it would satisfy the 90% case.

If it does, perhaps something dumber, like this could work?

```yaml
- repo: https://github.com/pre-commit/run-prebuilt-binary
  hooks:
    - id: download-and-run-binary
      alias: my-program
      args:
        - --target=https://github.com/user/my-program/releases/download/v1.0.0/my-program-v1.0.0-x86_64-apple-darwin.tar.xz
    
    - id: download-and-run-binary
      alias: another-program
      args:
        - --target=https://github.com/user/another-program/releases/download/v1.1.0/another-program-v1.1.0-x86_64-unknown-linux.tar.gz
```

I haven't fully thought this last suggestion through, but the idea of "just" pointing at the binary you want to run would be sufficient for my use-case (and I suspect a lot of others').

**Comment by sondrelg** [comment]
On reflection I guess my final suggestion above would also require specifying multiple targets - it's the same solution as the additional dependencies one, except the implementation would be separate from pre-commit itself.

**Comment by mattyclarkson** [comment]
Joining late, but we have solved this in the following way.

We implemented a Python CLI that can download a binary, check it against a subresource integrity and then execute the binary with some arguments.

We use it to grab pre-built binaries built in the CI and pushed into the [GitLab generic package registry](https://docs.gitlab.com/ee/user/packages/generic_packages/) or [Artifactory](https://jfrog.com/artifactory/). It would equally work with binaries published to a GitHub release.

We mainly do this for Rust and Go to prevent users having to have those SDKs installed (this was before we downloaded the Go/Rust SDK automatically). It can still save time for the user setting up the virtual environment as they just do the download rather than the compile.

It uses templating on the URL to find the correct binary for the users platform. `{os.go}` it the Go representation of the operating system (`linux`, `windows`, `darwin`), etc.

```yaml
- id: shfmt
  name: shfmt
  description: Shell source code formatter
  language: python
  language_version: python3
  entry: |
    execute
      -i linux/amd64:sha256-GpEiFJdPTtSzE3lfKsyRYP6Llc9wbYOzOcFgxn1xOoU=
      -i linux/arm64:sha256-yEOBMwjCKip8uNH5fP2RJXtGHyH/N+57MqTsi7KYa0M=
      -i darwin/amd64:sha256-NpEHmb2pSTY/oHhbUkuxTjFJRfFQVBOn4nFMqkAdHjs=
      -i darwin/arm64:sha256-1OHOgnd93nRICXrYY5YuXPxX/jSE4kyaurGtNOCSBw8=
      -i windows/amd64:sha256-nPkDO6xkd4GK12LhPMHavJXr6wONuBGr28F1tO997Sk=
      -i windows/arm64:sha256-F1vg6ayWmCpo7U1HtakCqDhdG2Ahsiy4kcLJA/pCrSE=
      --cache={file.parent}/.cache
      --url https://gitlab.gpu.arm.com/api/v4/projects/109/packages/generic/shfmt/3.6.0/shfmt.{os.go}-{cpu.go}
      --header "PRIVATE_TOKEN: {env.GITLAB_API_TOKEN}"
      --header "JOB_TOKEN: {env.CI_JOB_TOKEN}" -- -w -s
  types:
    - shell
  stages:
    - commit
    - push
    - manual
```

Not sure we need this as a new language in the core of `pre-commit`. We could create a new tool in `pre-commit-hooks` (or just a separate tool?) and use it as so:

```yaml
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.5.0
    hooks:
      id: download-and-execute
      args:
        - -i linux/amd64:sha256-GpEiFJdPTtSzE3lfKsyRYP6Llc9wbYOzOcFgxn1xOoU=
        - ...etc...
```

**Comment by asottile** [comment]
downloading at runtime is problematic as the cache is intended to be immutable

**Comment by sondrelg** [comment]
If so, the conclusion seems to be that the pre-commit model is generally incompatible with a "download" language. I think it would be fair to close the issue if you don't see an obvious way forward from here :+1: 
 
In case it's helpful for anyone else: the way I solved the not-wanting-to-build-my-hook-from-source issue for a rust pre-commit hook I was writing, was to simply build and publish python wheels for it. With [maturin](https://github.com/PyO3/maturin), building a wheel for a Rust pre-commit hook is as simple as `maturin build -b "bin" --release`. Full example [here](https://github.com/sondrelg/printf-log-formatter).

**Comment by asottile** [comment]
> If so, the conclusion seems to be that the pre-commit model is generally incompatible

no?

**Comment by dudicoco** [comment]
> downloading at runtime is problematic as the cache is intended to be immutable

Isn't that exactly what https://github.com/shellcheck-py/shellcheck-py does?

**Comment by dudicoco** [comment]
Ahh actually it seems that it does it as part of the setup process

**Comment by gilesw** [comment]
It does seem like setting up binary dependencies is a problem that a few different people have hit. I ended up using asdf and a bash script for developers and pre-commit-ci-lite with github actions. Would be great to have asdf support built in.

**Comment by mattyclarkson** [comment]
We're looking to respin our internal downloader into a PR, unless someone else gets to this first.

Something along the lines of:

```- id: shfmt
  name: shfmt
  description: Shell source code formatter
  language: download
  additional_dependencies:
    - |-
    linux/amd64
    sha256-GpEiFJdPTtSzE3lfKsyRYP6Llc9wbYOzOcFgxn1xOoU=
    https://gitlab.com/api/v4/projects/109/packages/generic/shfmt/3.6.0/shfmt.linux-amd64
    - |-
    linux/arm64
    sha256-yEOBMwjCKip8uNH5fP2RJXtGHyH/N+57MqTsi7KYa0M=
    https://gitlab.com/api/v4/projects/109/packages/generic/shfmt/3.6.0/shfmt.linux-arm64
  types:
    - shell
  stages:
    - commit
    - push
    - manual
```

A few thoughts:

* Will use the [Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) standard to describe the checksum for the downloaded file
* Splitting the additional dependencies on newlines to make the hook description more readable in the `.pre-commit-hooks.yaml`.
    * Not sure how this will play with the output on the CLI when the environment is being created 🤷 
    * We can split on `:` like other tools, if needed
* Authentication will be through `.netrc`.
* Unpacking of archives could be done with [shutil.unpack_archive](https://docs.python.org/3/library/shutil.html#shutil.unpack_archive)
    * It would be possible to add a `pre-commit` [entry point](https://docs.python.org/3/library/importlib.metadata.html#entry-points) that allows plugins to register extra unpacking routines above the default Python 3 `xz`, `bz`, `gzip` and `zip` ones.
    * We likely won't implement unpacking in the first PR

**Comment by asottile** [comment]
actionlint-py + shellcheck-py seems to work for other pre-commit.ci customers

**Comment by louwers** [comment]
Keeping an eye on actionlint-py, but it does not seem to support my CPU (Apple ARM).

**Comment by asottile** [comment]
there's arm64 binaries -- send a PR to them

**Comment by mattyclarkson** [comment]
As an update: we're rolling through the implementation of this. We've realised that we need an extra line to the additional dependencies to specify when the binary should be stored so that we have a consistent binary name for the `entry` key. Hoping to get something up soon™️.

```
  additional_dependencies:
    - |-
    linux/amd64
    sha256-GpEiFJdPTtSzE3lfKsyRYP6Llc9wbYOzOcFgxn1xOoU=
    https://gitlab.com/api/v4/projects/109/packages/generic/shfmt/3.6.0/shfmt.linux-amd64
    bin/shfmt
```

**Comment by scop** [comment]
Looks like there's quite a bit of movement in this area currently. Anyway I filed https://github.com/pre-commit/pre-commit.com/pull/904 which wasn't accepted but was asked to chime in here so here goes:

https://github.com/scop/wrun implements a generic "download launcher" with caching and archive support, not specific to pre-commit. It does work with pre-commit too though, just not with pre-commit.ci as it stands, and perhaps not fully the way pre-commit's languages are intended to work.

It might contain some ideas that a pre-commit specific implementation could make use of.

One comment about the subresource identity idea here: I considered that for wrun as well, but stuck with the "regular" hex digests with algorithm prefix. This is because these downloads are not really "subresources" in my opinion, and because IME hex digests are vastly more common than base64 ones in this problem space. Using base64 would mean one would practically need to generate or convert the digests separately instead of being able to use upstream provided ones as is. It might require some custom / not that common tooling as well or a few more hops, as the standard `sha256sum` and friends family of tools would not be applicable.

**Comment by Timmmm** [comment]
I just created [a simple Rust hook](https://github.com/timmmm/custom_regex_lint) and wondered how `pre-commit` was setting up Rust if a user doesn't have it. Turns out it installs it to `$HOME` in the normal way. Probably the only sensible option but I can guarantee if I introduce this into my company people are going to complain "why has something installed Rust when I didn't want it to"?

Anyway, a workaround I've used previously is to distribute binaries via Pip. Yes it sounds mad, but Pip is pretty much universally available and works quite well for distributing Rust programs. Here's [an example](https://github.com/timmmm/maxtime) - you can `pip install maxtime` and nobody is any the wiser that it's using Rust (some people are weirdly sensitive about that).

So, for Rust programs at least, can I:

1. Publish the binary to PyPI, as with `maxtime`.
2. Create a trivial Python wrapper hook that just lists `maxtime` as a dependency in `pyproject.toml`.

Is there any reason that wouldn't work? Will that all be cached?

**Comment by asottile** [comment]
> "why has something installed Rust when I didn't want it to"?

they would never know, it is not exposed outside of the pre-commit cache.  it's not installed to their homedir at all you must be confused

**Comment by gytis-ivaskevicius** [comment]
Fetching prebuilt binaries is not that simple, often one needs to support 3 operating systems and 2 architectures but maybe someone will find this helpful:
We use https://github.com/numtide/devshell to add required binaries to the `$PATH` and `direnv` to auto load devshell once we cd into the directory. Also, it is super simple to configure a little MOTD to help users get started. Example:
```
🔨 Welcome to lace-ops

[cloud]

  aws                  - Unified tool to manage your AWS services
  eksctl               - CLI for Amazon EKS
  opentofu             - Tool for building, changing, and versioning infrastructure
  sops                 - Simple and flexible tool for managing secrets
  terragrunt           - Thin wrapper for Terraform that supports locking for Terraform state and enforces best practices

[direct access]

  k9s-eu-central-1
  k9s-us-east-1
  k9s-us-east-2

[general commands]

  jq                   - Lightweight and flexible command-line JSON processor
  just                 - Handy way to save and run project-specific commands
  lace                 - Lace
  menu                 - prints this menu
  treefmt              - one CLI to format the code tree
  yq-go                - Portable command-line YAML processor

[kubernetes]

  k9s                  - Kubernetes CLI To Manage Your Clusters In Style
  kubectl              - Kubernetes CLI
  kubernetes-helm      - Package manager for kubernetes

[terraform]

  tf-ci
  tf-dev-us-east-1
  tf-global
  tf-prod-eu-central-1
  tf-prod-us-east-2
  ```

**Comment by asottile** [comment]
@gytis-ivaskevicius this is not an advertising space for off topic tools

**Comment by awoimbee** [comment]
Honestly I don't see how shipping binary dependencies can work without a hook postinstall script.
A simple 'download' language would not support setting permissions, unzipping, etc. For example, here is one of my Makefile target:

```Make
ALLOY := $(BIN)/alloy-$(VERSION_ALLOY)-$(OS)-$(ARCH_BIS)
$(ALLOY): | $(TMP) $(BIN)
	curl -L https://github.com/grafana/alloy/releases/download/$(VERSION_ALLOY)/alloy-$(OS)-$(ARCH_BIS).zip | funzip > $(TMP)/alloy
	chmod +x $(TMP)/alloy
	$(TMP)/alloy --help >/dev/null
	mv $(TMP)/alloy $@
```

**Comment by c0m1c5an5** [comment]
[wrun](https://github.com/scop/wrun) is now archived and suggests using [mise](https://github.com/jdx/mise) or [aqua](https://github.com/aquaproj/aqua) instead. I found this thread during a research on previous issues before creating a new issue for adding the `aqua` language and I am willing to work on a PR for this language.

Edit: After spending some time with aqua both locally and in CI, I no longer see a reason for adding support as a language specifically, as regular aqua install + sysyem/python language achieves most requirements. I guess the only advantage of having it as a language would be pre-commit.ci support.

**Comment by c0m1c5an5** [comment]
Aqua is quite reliable and easy to use.
My suggestion for this case is:

```yaml
# .pre-commit-config.yaml
  - repo: local
    hooks:
      - id: aqua-example
        name: Example aquaproj/aqua hook
        entry: aqua
        args:
          - exec
          - --
          - yq
          - --help
        language: python
        additional_dependencies:
          - aquaproj-aqua==2.60.0.post0
```

```yaml
# aqua.yaml
---
# yaml-language-server: $schema=https://raw.githubusercontent.com/aquaproj/aqua/main/json-schema/aqua-yaml.json
# aqua - Declarative CLI Version Manager
# https://aquaproj.github.io/
checksum:
  enabled: true
  require_checksum: true
  supported_envs:
    - all
registries:
- type: standard
  ref: v4.524.0 # renovate: depName=aquaproj/aqua-registry
packages:
  - name: mikefarah/yq@v4.50.1
```


```bash
$ pre-commit run --all-files -v aqua-example
Example aquaproj/aqua hook...............................................Passed
- hook id: aqua-example
- duration: 0.1s

yq is a portable command-line data file processor (https://github.com/mikefarah/yq/) 
See https://mikefarah.gitbook.io/yq/ for detailed documentation and examples.

## Evaluate Sequence ##
This command iterates over each yaml document from each given file, applies the 
expression and prints the result
```

**Comment by asottile** [comment]
that's not really relevant for this discussion

**Comment by c0m1c5an5** [comment]
> that's not really relevant for this discussion

At this point after 6 years of this not being implemented, it is relevant.
At least we can use this solution while waiting for someone to re-implement everything aqua does in python.

**Comment by asottile** [comment]
there's already conda if you want that.  no it's not relevant here
